app/controllers/glossary_terms_controller.rb

```diff
@@ -99,28 +98,55 @@ def show_past
 
   # --------- Other private methods
 
-  def image_args
-    Rails.env.test? ? permit_upload_image_param : strong_upload_image_param
+  def init_image_form_instance_variables
+    @copyright_holder ||= @user.name
+    @copyright_year ||= Time.now.utc.year
+    @upload_license_id ||= @user.license_id
+    # rubocop:disable Naming/MemoizedInstanceVariableName
+    # RuboCop wants to name the ivar "@update_image_form_instance_variables"
+    @licenses ||= License.current_names_and_ids(@user.license)
+    # rubocop:enable Naming/MemoizedInstanceVariableName
   end
 
-  def strong_upload_image_param
-    {
-      copyright_holder: params[:copyright_holder],
-      when: Time.local(params[:date][:copyright_year]).utc,
-      license: License.safe_find(params[:upload][:license_id]),
-      user: @user,
-      image: params[:glossary_term][:upload_image]
-    }
+  def reload_form(form)
+    add_glossary_term_error_messages_to_flash
+    init_image_form_instance_variables
+    render(form)
   end
 
-  def permit_upload_image_param
-    args = strong_upload_image_param
-    # Remove "permitted: false" from this param so that model can mass assign it
-    args[:image] = params[:glossary_term][:upload_image][:image]
-    args
+  def add_glossary_term_error_messages_to_flash
+    @glossary_term.errors.messages.each_value do |val|
+      # flash_error takes a string; val is an array of size 1, e.g. ["message"]
+      flash_error(val.first)
+    end
+  end
+
+  # Process any image together with @glossary_term,
+  # returning truthy if neither fails
+  # They must be processed together to correctly validate GlossaryTerm and
+  # allow backing out Image if GlossaryTerm is invalid
+  def image_and_term_saves_smooth?
+    # If no upload file specified, only issue is the term
+    return @glossary_term.save unless upload?
+
+    # return false if image processing fails
+    return unless (saved_image = process_upload(image_args))
+
+    @glossary_term.add_image(saved_image)
+    return if @glossary_term.save # happy path
+
+    # term failed, so clean up the orphaned (unassociated) image
+    # and its flash notice ("Successfully uploaded image ...")
+    saved_image.try(:destroy)
+    flash_clear
+    false
+  end
+
+  def upload?
```

Comment:
Very minor point: but would you mind changing the name to "upload_present?" or equivalent so that it is more readily apparent that this method is not a verb (meaning an action actually occurs)?

Reply:
Very good point! I will use `upload_specified?`